File: app/src/main/java/org/ole/planet/myplanet/ui/teams/leaderboard/TeamLeaderboardCalculator.kt
package org.ole.planet.myplanet.ui.teams.leaderboard

import com.google.gson.JsonObject
import java.util.Calendar
import org.ole.planet.myplanet.model.RealmUser
import org.ole.planet.myplanet.model.TeamLeaderboardEntry
import org.ole.planet.myplanet.repository.JoinedMemberData

object TeamLeaderboardCalculator {
    fun build(
        members: List<JoinedMemberData>,
        courseIds: List<String>,
        progressByUser: Map<String, Map<String?, JsonObject>>,
        surveyCompletionTimestampsByUser: Map<String, List<Long>> = emptyMap(),
        surveysTotal: Int = 0,
        currentUserId: String? = null,
        periodStart: Long? = null
    ): List<TeamLeaderboardEntry> {
        val entries = members.mapNotNull { member ->
            val userId = member.user.id ?: return@mapNotNull null
            val progress = progressByUser[userId].orEmpty()
            var coursesCompleted = 0
            for (courseId in courseIds) {
                val courseProgress = progress[courseId] ?: continue
                val max = courseProgress.get("max")?.asInt ?: 0
                val current = (courseProgress.get("current")?.asInt ?: 0).coerceAtMost(max)
                val completedAt = courseProgress.get("completedAt")?.asLong ?: 0L
                val isCompleted = max > 0 && current >= max &&
                    (periodStart == null || completedAt >= periodStart)
                if (isCompleted) coursesCompleted++
            }
            val surveyTimestamps = surveyCompletionTimestampsByUser[userId].orEmpty()
            val surveysCompleted = surveyTimestamps.count { periodStart == null || it >= periodStart }
            TeamLeaderboardEntry(
                visitInfo = member,
                displayName = displayName(member.user),
                coursesCompleted = coursesCompleted,
                coursesTotal = courseIds.size,
                surveysCompleted = surveysCompleted,
                surveysTotal = surveysTotal,
                isCurrentUser = !currentUserId.isNullOrEmpty() && userId == currentUserId
            )
        }
        return entries.sortedWith(
            compareByDescending<TeamLeaderboardEntry> { it.coursesCompleted }
                .thenByDescending { it.surveysCompleted }
        )
    }

    fun startOfCurrentMonth(nowMillis: Long): Long {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = nowMillis
        calendar.set(Calendar.DAY_OF_MONTH, 1)
        calendar.set(Calendar.HOUR_OF_DAY, 0)
        calendar.set(Calendar.MINUTE, 0)
        calendar.set(Calendar.SECOND, 0)
        calendar.set(Calendar.MILLISECOND, 0)
        return calendar.timeInMillis
    }

    private fun displayName(member: RealmUser): String {
        val combined = "${member.firstName.orEmpty()} ${member.lastName.orEmpty()}".trim()
        return combined.ifBlank { member.name.orEmpty() }
    }
}

File: flutter/lib/ui/teams/leaderboard/team_leaderboard_calculator.dart
import '../../../repository/progress_repository.dart'
    show CourseProgressSummary;

/// Port of `model/TeamLeaderboardEntry.kt`. One row per team member in the
/// leaderboard: their display name, courses completed, surveys completed,
/// and whether they are the current user (for highlighting).
class TeamLeaderboardEntry {
  const TeamLeaderboardEntry({
    required this.userId,
    required this.displayName,
    required this.coursesCompleted,
    required this.coursesTotal,
    required this.surveysCompleted,
    required this.surveysTotal,
    required this.isCurrentUser,
    required this.visitCount,
    this.userImage,
  });

  final String userId;
  final String displayName;
  final int coursesCompleted;
  final int coursesTotal;
  final int surveysCompleted;
  final int surveysTotal;
  final bool isCurrentUser;
  final int visitCount;
  final String? userImage;
}

/// Port of `TeamLeaderboardCalculator`. Pure logic: given members, course
/// progress, and survey completion timestamps, build the ranked list sorted
/// by courses completed then surveys completed (both descending).
class TeamLeaderboardCalculator {
  const TeamLeaderboardCalculator._();

  static List<TeamLeaderboardEntry> build({
    required List<MemberWithProgress> members,
    required List<String> courseIds,
    required Map<String, CourseProgressSummary> Function(String userId)
    progressForUser,
    required List<int> Function(String userId) surveyTimestampsForUser,
    required int surveysTotal,
    required String? currentUserId,
    int? periodStart,
  }) {
    final entries = <TeamLeaderboardEntry>[];
    for (final member in members) {
      final progress = progressForUser(member.userId);
      var coursesCompleted = 0;
      for (final courseId in courseIds) {
        final cp = progress[courseId];
        if (cp == null) continue;
        final isCompleted = cp.max > 0 && cp.current >= cp.max;
        if (isCompleted) coursesCompleted++;
      }
      final timestamps = surveyTimestampsForUser(member.userId);
      final surveysCompleted = periodStart == null
          ? timestamps.length
          : timestamps.where((t) => t >= periodStart).length;
      entries.add(
        TeamLeaderboardEntry(
          userId: member.userId,
          displayName: member.displayName,
          coursesCompleted: coursesCompleted,
          coursesTotal: courseIds.length,
          surveysCompleted: surveysCompleted,
          surveysTotal: surveysTotal,
          isCurrentUser:
              currentUserId != null && member.userId == currentUserId,
          visitCount: member.visitCount,
          userImage: member.userImage,
        ),
      );
    }
    entries.sort((a, b) {
      final byCourses = b.coursesCompleted.compareTo(a.coursesCompleted);
      if (byCourses != 0) return byCourses;
      final bySurveys = b.surveysCompleted.compareTo(a.surveysCompleted);
      if (bySurveys != 0) return bySurveys;
      // Members arrive through a `Set` and `List.sort` is not stable, so
      // without a total order two members with identical scores could swap
      // rank between two loads of unchanged data. Name then id, so the
      // ranking is reproducible.
      final byName = a.displayName.toLowerCase().compareTo(
        b.displayName.toLowerCase(),
      );
      return byName != 0 ? byName : a.userId.compareTo(b.userId);
    });
    return entries;
  }

  /// Port of `startOfCurrentMonth` — the first millisecond of the current
  /// month in UTC, for the "this month" period filter.
  static int startOfCurrentMonth(int nowMillis) {
    final now = DateTime.fromMillisecondsSinceEpoch(nowMillis, isUtc: true);
    return DateTime.utc(now.year, now.month, 1).millisecondsSinceEpoch;
  }
}

/// A member with the fields the leaderboard needs, resolved from the users
/// table and team-log visit counts.
class MemberWithProgress {
  const MemberWithProgress({
    required this.userId,
    required this.displayName,
    required this.visitCount,
    this.userImage,
  });

  final String userId;
  final String displayName;
  final int visitCount;
  final String? userImage;
}
